### DISCUSSION TITLE
MPDisplay:  a CROSSPLATFORM Display and Event driver for CPython, CircuitPython and MicroPython

### DISCUSSION THREAD

**Discussion by bdbarnett** [micropython/micropython#15261] [Show and tell]
[MPDisplay](https://github.com/bdbarnett/mpdisplay) is a display and event driver that works on CPython, CircuitPython and MicroPython.  It enables you to develop your app's UI on your deskop and the code immediately transfers to your microcontroller!

- It provides a full graphics library and works with input devices including touchscreens, encoders and keypads on microcontrollers as well as mice, scroll wheels and keyboards on desktop operating systems.
- It supports MicroPython on Unix and on microcontrollers.
- Out of the box it works with SPI and i80 buses (tested thoroughly on ESP32 family and RP2040), and it will work on top of bus drivers provided by others, such as the bus drivers in @kdschlosser's lvgl_micropython bindings or the bus drivers in CircuitPython, namely FourWire, ParallelBus and DotClockFrameBuffer.
- If you're running CircuitPython on RP2040, it can make your board look like a webcam to your Unix desktop OS, sending what would be on a display as a video stream to your desktop's camera app or videoconferencing software.  Even works on Chromebooks!
- It can function under a GUI library, such as LVGL or @peterhinch's MicroPython-Touch and Nano-GUIs **without modification**.
- It works with (and includes) all of @russhughes's examples from his [st7789py_mpy](https://github.com/russhughes/st7789py_mpy) with only 2 simple modifications.
- It has classes and functions for multiple types of fonts, including those from @peterhinch and @russhughes, and can open RGB565 encoded .BMP files.
- It is designed to be a platform to build GUI libraries on, but is very capable on it's own.
- It has 3 super easy to use installers:  a bash script for Unix, a Windows PowerShell script for Windows and a .json installer for mip on Micropython.
- On most Unix environments, including WSL, you can be running your first example within 60 seconds of launching the installer.
- It comes with LOTS of example code that runs on all the platforms.
- The documentation is still a work in progress, but it already details what it is, what it does, and how to get it.  There are examples for every feature it provides, so you should be up and running your own code in no time.

Please take 5 minutes and install it on your desktop / laptop and see what you think.  If you like it, give it a star.  If you can help improve it, submit a pull request or message me on its discussion board.  I appreciate your time and feedback!

https://github.com/bdbarnett/mpdisplay

**Comment by peterhinch** [comment]
Ingenious!

I attempted to run `nano_gui_simpletest.py` and got errors - I've raised an issue. The micropython-touch demo `active.py` runs but is not touch sensitive.

The calculator demo sort-of-works in that I can enter numbers but there seem to be missing buttons - e.g. `+`, `-` and `/`. I can raise an issue with a screenshot if you wish.

**Reply by bdbarnett** [reply]
Thank you very much sir!

I replied to the issue and think we're OK there.  I'm sorry you're finding bugs already.  I've refactored too much code in the last 2 weeks.

Are the buttons completely missing in calculator, or just the labels?  I'm not having the same issue on CPython under WSL, micropython under WSL or CPython under Windows

**Comment by jonnor** [comment]
Looks great! I especially appreciate that host/PC is a first-class platform, developing in a portable manner is really the way to go for modern embedded development imo.

**Comment by easytarget** [comment]
> If you're running CircuitPython on RP2040, it can make your board look like a webcam to your Unix desktop OS, sending what would be on a display as a video stream to your desktop's camera app or videoconferencing software. Even works on Chromebooks!

Sublime!  what an astoundingly good idea!
